File: backend/app/core/upgrades/stat_modifiers/dash_cooldown_1.py
from ..upgrade_implementation import UpgradeImplementation, UpgradeAppliers


class DashCooldown1(UpgradeImplementation):
    id = 'dash_cooldown_1'
    name = 'Swift Recovery'
    description = '-5% dash cooldown'
    rarity = 'rare'
    type = 'stat_modifier'
    cost = 10

    def can_apply(self, state: dict) -> bool:
        return True

    def apply(self, appliers: UpgradeAppliers) -> None:
        value = -0.05
        appliers.modifiers.add_multiplier('player', 'dashCooldown', value)

    def validate(self, recorded_stats: dict) -> bool:
        return True

File: frontend/src/game/upgrades/stat_modifiers/dash_cooldown_1.ts
import { Upgrade, type UpgradeDef } from '../Upgrade'
import { UpgradeTargetID, UpgradeStatID, RarityID, UpgradeTypeID } from '../../data/ID'

export const DashCooldown1Def: UpgradeDef = {
  id: "dash_cooldown_1",
  name: "Swift Recovery",
  description: "-5% dash cooldown",
  rarity: RarityID.Rare,
  upgradeType: UpgradeTypeID.StatModifier,
  cost: 10,
  targetClass: UpgradeTargetID.Player,
  fieldInTargetClass: UpgradeStatID.DashCooldown,
  value: -0.05,
  isMultiplier: true,
  stackable: true,
  maxStacks: 5,
  dependentOn: [{ ids: ["dash_ability"] }],
}

export class DashCooldown1 extends Upgrade {}
